File: two-pointers/valid-palindrome-ii.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
  public:
    bool validPalindrome(string &s) {
      int low = 0, high = s.size() - 1;

      while (high > low) {
        if (s[low] != s[high])
          return isPalindrome(s, low + 1, high) || isPalindrome(s, low, high - 1);

        ++low;
        --high;
      }

      return true;
    }

  private:
    bool isPalindrome(string &s, int start, int end) {
      while (end > start)
        if (s[start++] != s[end--])
          return false;

      return true;
    }
};

File: two-pointers/valid_palindrome_ii.py
class Solution:
    def validPalindrome(self, s: str) -> bool:
        def isPalindrome(s: str) -> bool:
            return s == s[::-1]

        left, right = 0, len(s) - 1

        while left < right:
            if s[left] != s[right]:
                del_left = s[left + 1:right + 1]
                del_right = s[left:right]

                return isPalindrome(del_left) or isPalindrome(del_right)

            left += 1
            right -= 1

        return True